### DISCUSSION TITLE
Blastn Scientific Names Output

### DISCUSSION THREAD

**Discussion by skreling** [sr320/course-fish546-2021#86] [Q&A] [answered]
tl;dr: blastn is working, but I can't figure out how to incorporate the taxdb into blast's makedb to get an output with full scientific names.  

I've finally gotten Blastn to work on most of my samples (think there is just an issue with one of original files so I need to redownload it), but I want to create a tab (format 6) output that gives scientific names and common names. In order to do that it requires this database (ftp://ftp.ncbi.nlm.nih.gov/blast/db/taxdb.tar.gz) to pull the scientific names from. However, I can't seem to get it to work. I've tried the suggested putting the untarred file in the same directory as the database, but can't figure out which blast makedb argument to incorporate it in. Maybe I'm just not understanding what needs to happen, but this forum (https://www.biostars.org/p/76551/) has been a bit of a help and I think a lot of it is just going over my head.

Here (https://www.ncbi.nlm.nih.gov/books/NBK279690/#CmdLineAppsManual.Quick_start) is also the manual for blast and I've looked at a good portion of the related content and it still hasn't really helped me get what I want

**Comment by kubu4** [comment]
When you have a sec, can you please share code you ran to make the database and run BLAST, as well as sample of resulting output file?

**Reply by skreling** [reply]
Here's an example output:
https://github.com/fish546-2021/Sam-Metabarcoding/blob/main/Results/Bat/1_S1_L001_R1_001_reference_blastn.tab
You'll see that right now I just have the reference file but not the taxadb file in there because I can't figure out where it goes

Here is the code:
https://github.com/fish546-2021/Sam-Metabarcoding/blob/main/Scripts/01_Bat_Blast.ipynb

Also it's Sunday! Feel free to ignore this and enjoy your weekend!

**Comment by kubu4** [comment]
Your notebook shows you using this file to build the BLAST database with this file:

`reference_fasta.fas`

Can you please provide a link to this file (if you need a URL, please upload to Gannet and then go here with your web browser to find the file: https://gannet.fish.washington.edu/fish546_2021/skreling/)?

Can you also provide us some background on where you got this file and how it was generated?

Also, why are you using this file? 

The answers to the above questions might allow us to negate most of the stuff below!

Looking at the FastA contents in your notebook, you'll be creating a custom BLAST database. In order to create a custom BLAST database which will allow you to assign taxonomies in the BLAST outputs, you need to build the BLAST database using a custom taxonomy mapping file for your FastA.

To do that, you need a mapping file that would look like this (this is just an example and cannot be used for anything):

```
GBACU292-12|Tachysphex pompiliformis|COI-5P|MH611210 68287
GBACU2923-13|Tachysphex obscuripennis|COI-5P|MH611205 2382161
GBACU2924-13|Diodontus wahisi|COI-5P|MH608492 68287
GBACU2925-13|Diodontus wahisi|COI-5P|MH608953 382
```

The mapping file is formatted like this (space-delimited):

`fasta_sequence_header NCBI_taxonomy_node`

Are you ready for an exercise in file parsing? :)

Here are the steps you should follow:

1. Download and install [NCBI Entrez Direct](https://www.ncbi.nlm.nih.gov/books/NBK179288/). This is a tool that allows querying a variety of NCBI resources. It's required for the `ncbi-blast-2.11.0+/bin/get_species_taxids.sh`, which we'll get to in a bit. It also requires an active internet connection to use.

2. Test out the `ncbi-blast-2.11.0+/bin/get_species_taxids.sh` script using the first species name from your FastA (you will _need_ quotes around the gene species). You can ignore the `WARNING` in the output. Also, you're wanting the Taxid value.

```bash
./get_species_taxids.sh -n "Tachysphex pompiliformis"
```

```
WARNING: Redundant -db 'taxonomy' argument
 
Taxid : 1889563 
 rank : species 
 division : bees 
 scientific name : Tachysphex pompiliformis 
 common name : 

1 matche(s) found.
```

3. Start scripting/parsing (FYI, I'm happy to help with this, but figured you might like to give it a shot first). Here's a rough outline:

    1. Figure out a command to parse out the genus/species from FastA header lines. To keep things "simple", work only on a single header (e.g. ```header=">GBACU292-12|Tachysphex pompiliformis|COI-5P|MH611210"``` and then you can use `echo "${header}" | parsing-commands` for testing).
    2. Figure out a command to parse out the Taxid from the `./get_species_taxids.sh -n "Genus species" results.
    3. Figure out how to save the results of the two parsing tasks above to their own variables.
    4. Figure out how to pass the genus/species as an argument to the `./get_species_taxids.sh -n` "genus species"`.
    4. Figure out how to print each of the parsing results in the expected mapping file format. Hint: Use `printf`.
    5. Use the above knowledge to loop through FastA, like so:

```bash
while read -r line
do
  # If the line begins with ">", then parse it
  if [[ "${line}" = ">"* ]]; then
  parsing_command_stuff
  parsing_command_stuff
  fi
done < reference_fasta.fas >> tax-id_map.txt
```

Aaaaand, that's all there is to it! Easy, right? LOL. This is definitely complicated, so please do not hesitate to ask more questions! I recommend starting a new discussion for each question/problem you might encounter.

**Reply by skreling** [reply]
Ok thank you. I think the issue is it has to be an NCBI database and what I pulled was from BOLD. I'll try a NCBI database and see if that resolves the problem first before messing with the other reference file since that isn't necessarily what I have to be using!  

Thanks!

**Reply by skreling** [reply]
I tried parsing, and got to a theoretical level of what I wanted to do and then had no idea of what to do after that. My idea since the file is all formatted the same way was to essentially extract the string after there first "|" and before the second "|".


I got close with this, but it gave me COI-5P instead of the genus species.

```!echo "${header}" | sed -e 's/.*|\(.*\)|.*/\1/' #close but too far over?```

**Comment by kubu4** [comment]
Please post the results of the `head` command for the file you're working with. Also, the BLASTn command you used would be helpful.

**Reply by skreling** [reply]
Maybe I should clarify things a bit. I ended up working with the original reference file, just because it was a bit more human-readable than the ncbi alternatives and I'm all about trying to make things ever so slightly easier right now. 

Here's the Jupyter NB I've been working from. You can see where I was trying to parse just the species id out of the header, but that didn't really work well. Essentially stuck at part ii. of your workflow above if that makes sense. 

https://github.com/fish546-2021/Sam-Metabarcoding/blob/main/Scripts/.ipynb_checkpoints/01_Bat_Blast-checkpoint.ipynb

**Comment by kubu4** [comment, marked as answer]
Just popping this in here for posterity. Per class today, @sr320 recommended editing your reference FastA file (BOLD?) to replace space between `genus species` in FastA headers with a character (`_`).

You suggested possibly using `sed` which would work perfectly!
